Fix Arduino cycle hickup

joluxer opened the pull request:
After 71 minutes, the PLC cycle is broken and runs too fast or not at all for several millseconds to seconds. At lease one time the Arduino came to complete stand-still.

The reason is at evaluation of the micro-seconds counter. This evaluation fails on counter overrun. A number of 2^32 micro-seconds equals 71,6 minutes, where the symptoms occur.

joluxer committed:
Introduce file attributes to git for cross-platform development.

On different execution and development platforms are different EOL
characters. And even some files shall not at all converted. This can
be controlled by .gitattributes. Starting with a safe default.

joluxer committed:
Close Arduino C file after reading in.

joluxer committed:
Text attribute for python sources.

joluxer committed:
Rearrange timing checks for reproducible time deltas also on wrap-around.

Microseconds timer wraps around approx. every 71 minutes, then some
hickup happens with one too short cycle.

b-uttley committed:
Changes for ESP32C-6 support and analog I/O problems ESP32 and Blackpill

joluxer commented:
Here are some pictures from my oscilloscope, that show the problem.

This is the test sketch, running every 95 ms:
![test-circuit](https://github.com/user-attachments/assets/654bdc45-ac5b-4c08-bd41-ec7285a86440)

This is the trigger code for the oscilloscope:
![test-trigger](https://github.com/user-attachments/assets/da22146a-adf0-455d-b41d-a9d8a94a31a7)

A normal PLC cycle before repair, oscilloscope triggered by the LED:

![normal-cycle](https://github.com/user-attachments/assets/61d7a754-11e9-4af0-a7db-c32a96162db0)

A couple of broken cycles, oscilloscope triggered by the trigger code shown above:

![broken-cycle-1](https://github.com/user-attachments/assets/727c09b7-839e-4971-9e2e-705a48140a0a)
![broken-cycle-2](https://github.com/user-attachments/assets/5d0cfba6-0fd9-40de-9885-5d10bc7f41f2)
![broken-cycle-3](https://github.com/user-attachments/assets/35b768cc-f964-4411-a769-132f4957808a)

![broken-cycle-4](https://github.com/user-attachments/assets/7342869d-ae80-436c-b82e-2a0597a06088)

Zoomed in on previous image:

![broken-cycle-4zoom](https://github.com/user-attachments/assets/5c3915fd-c469-4b67-9279-858aa090a546)

After repair, these symptoms do not longer occur.

A normal PLC cycle after scheduling repair, triggered by the LED:

![normal-cycle-new-sched](https://github.com/user-attachments/assets/f44360e6-71bb-4336-8475-46b999dad41a)

And here is the counter overflow cycle, triggered by the trigger code:

![counter-overflow-cycle-new-sched](https://github.com/user-attachments/assets/655fbb82-d534-4605-b861-27d28faa1539)

No more disturbances.

The Arduino does not stop, but continues to run smooth and constant.

joluxer commented:
As practiced before, the PR for the super-project comes in after this merge for easier integration and easier PR handling (smile).

Best regards, Johnny (aka J'Lo).

joluxer commented:
Here is little picture of my test setup:
![OpenPLC-ESP32-WROOM-cycle-hickup-test-setup](https://github.com/user-attachments/assets/4f72724c-e362-4f9b-a509-87b4f650e4aa)

joluxer committed:
Fetch host information about PATH and CPU and log it along with build output.

joluxer committed:
Compact the Popen call for easier maintenance.

johannes-lode commented:
I rebased this PR to PR#8, so PR#8 should be merged first, then this here.

joluxer committed:
Implement thread-safe text output to the GUI and build cache cleaning.

Thread-safe text output uses a communication queue and delegates all
text output widget manupulation to the GUI thread and preserves the
output order of the text, without extra buffering. Text buffering is
done in the widget by the system.

Build cache cleaning is implemented in several escalation stages of
using or cleaning the project build cache, reinstalling all libraries
or reinstalling all boards and libraries.

joluxer commented:
Please, stay tuned, I'll rebase this soon...

thiagoralves committed:
Fix crash when build folder cannot be cleaned

Windows sometimes prevents OpenPLC Editor to clean the build folder with a permission denied error. This fix does not work around that, but instead avoids the application to crash. Compilation could fail as a result of the build folder not being fully cleaned.

thiagoralves committed:
Update version.py

thiagoralves committed:
Fixed unicode errors on Windows

thiagoralves committed:
Update version.py

Commit:
Improved debugger connection with remote devices

joluxer committed:
WIP: archiving many tries of reducing GUI events during build output.

joluxer committed:
Using styled text control for more perfomance. Goiong to test on Darwin and Windows.

joluxer committed:
New text widget works reliably and performant without GUI congestion.

Commit:
Adding support for CONTROLLINO MICRO

Commit:
Update version.pyr

joluxer committed:
Rework build options into a ComboBox.

joluxer committed:
Rearrange COM port ComboBox to the transfer page.

joluxer committed:
Add board manager URL to hals.

thiagoralves committed:
Update version.py

thiagoralves committed:
Added support for Raspberry Pico 2 - including RISC V cores

thiagoralves committed:
Fix exception error when Windows blocks deletion of build cache files

thiagoralves committed:
Update version.py

joluxer committed:
WIP: while reworking builder for gradual build system update options.

joluxer committed:
WIP: Build sequence implemented and half tested.

joluxer committed:
WIP: add more Arduino environment checking after tests. More tests to come.

joluxer committed:
WIP: change for spaces-resilient call of external tools.

joluxer committed:
Clean up literal directories and tool paths.

joluxer committed:
Repair Arduino sketch extension defines.

joluxer committed:
Speed up core-update call.

joluxer committed:
Use multiple Arduino extension C files for Arduino build.

joluxer committed:
Log tools version in build.log.

joluxer committed:
Improve Windows support for running external tools, repair arduino_sketch.h.

joluxer committed:
Removing hassle around spaces in paths and file names.

thiagoralves committed:
Update version.py

joluxer committed:
Add Arduino Nano ESP32 and fix Arduino Giga Modbus TCP

joluxer committed:
Fix hidden characters in hals.json file
# Conflicts:
#	arduino/examples/Baremetal/hals.json

joluxer committed:
Update hals.json
# Conflicts:
#	arduino/examples/Baremetal/hals.json

joluxer committed:
Fix compilation error on some platforms when the program uses global variables.

joluxer committed:
Update version.py
# Conflicts:
#	version.py

joluxer committed:
Fixes analog IO on Raspberry Pico
# Conflicts:
#	arduino/src/hal/rp2040pico.cpp

joluxer committed:
Resolve .DS_store conflict.

joluxer committed:
Resolve .DS_store conflict.

joluxer committed:
Resolve conflicts.

joluxer committed:
Introduce tested JSON tools for hals.json etc., repair spaces-in-path problem.

joluxer committed:
Improve tools for board and core removal to be more platform-agnostic.

joluxer committed:
Move hard-coded compiler flags to hals.json.

joluxer committed:
Check core update not when using build cache.

joluxer committed:
Record settings changes of ArduinoUploadDialog immediately without need for a "Save" button.

Add events for all inputs.

joluxer committed:
Repair and stabilize loading of default values for Arduino settings.

joluxer committed:
Repair build option comparison.

joluxer committed:
Repair board type changes and do some cosmetic layout change.

The setting of IO defaults was delayed by one change of board, now it
is in sync.

johannes-lode commented:
I've rebased this branch, to reflect the most recent improvements on the editor, regarding the build output congestion fix and the issues with the IO defaults.

thiagoralves commented:
This PR modifies 170 files, which do not seem to be related to the original problem. Would you mind separating the different issues into different PRs? This will make it easier to track changes over time in the project. Thanks!

thiagoralves commented:
Also, I've checked your commit that fixes the issue here: https://github.com/thiagoralves/editor/pull/7/commits/6d3285c9252304ec6c80bee92fb8ae9107cd6236

Correct me if I'm wrong, but it seems to fix just one part of the problem, but not the whole thing. For example, on a near wraparound condition, `last_run += scan_cycle;` can cause `last_run` to wraparound, making it appear much smaller than `micros()`, which will then cause multiple executions in quick succession of the `loop()` function before the next actual scan cycle period has elapsed (i.e. `if ((micros() - last_run) >= scan_cycle)` will be true until `micros()` also wraps around). Perhaps we will have to add a check to verify if the variable has wrapped around and then adjust accordingly

joluxer commented:
The mechnism of all the wraparound stuff is like follows:
In C/C++ (in contrast to Rust and Nim) an integer calculation, that does an overflow, does it for addition/increment in the same consistent way like for subtraction/decrement. And so it does on assembly and machine language level.

Thus the statement `last_run += scan_cycle;` wraps around - that is correct - and `last_run` will appear much smaller than `micros()`. But: the difference `((micros() - last_run)` will remain intact, because it does a wraparound/overflow in the opposite direction also correctly. 
This is defined by the underlaying data type. So if `last_run` becomes very small in the vicinity to the wraparound point and `micros()`is roughly in the same vicinity, the formulated difference maintains the correct positive value, that stands for the time difference on the wall clock. The order of magnitude to define vicinity is clearly defined by the half of the number space of the underlaying data type. That means, if the wall clock time difference of `((micros() - last_run)` is smaller that half of its data type, the the term always stands for the precise positive time difference. And the difference can only trigger once, as then the `last_run` value is set forward and it will trigger later. If the CPU takes too long between two checks, then two checks can trigger consecutively and if the CPU could not catch up for several cycles, then this happens too, but if the average CPU load is below 100%, this counter will catch up by multiple consecutive triggers, until all triggers are done.
So for our case of 32 bits of microseconds, the maximum time difference must be smaller than `~36` minutes (less than half of 73 minutes) to be calculated correctly using this wraparound-technique. Most of our timing will range in the order of 10~100 of milliseconds, sometimes seconds, so we are far away from the critical amount of time difference, where the calculation could fail. 

As an embedded systems engineer I've implemented already a lot of time triggered systems using this technique and it was the only working solution. All other variants, trying wraparound handling, testing conditions etc. failed sooner or later, when an interrupt based clock wrapped around during those tests. It might seem somewhat scary in the first, but if you hold it against the ALU definition of the underlying CPUs, if you watch in machine code or assembly code, if you simulate it on a CPU simulator, all this will proof, that this is reliable and correct. I'm confident, that this solution is robust up to the mentioned limits (in our case ~36 minutes of maximum sleep time).

If you still have more questions on this, please feel free to ask. I'm open to discuss this to answer all open questions.

If you are interested, I can provide you a simulation for that, implemented in Excel/LibreofficeCalc, where you can play around with those mechanisms. In this simulation you can set the speed of the `micros()` clock, the wraparound value for the data type and you can set the scan cycle.

Regards, Johnny

thiagoralves commented:
Hi Johnny, thanks for the detailed info. I put some more thought on this, and you're right! I was not worried about the micros() - last_run difference, I knew it would always output a positive integer over wrap arounds given that both variables are unsigned. My worry was if last_run wrapped around before micros(). If that happened, then indeed we would have a problem. After a quick GDB session I realized that this is not possible. The last_run variable is always behind micros(), as you can only increment it up to the point where it is equal micros(), never bigger. Therefore, micros() will always wrap around first, keeping the algorithm steady.

If anyone is interested in investigating this further, I created a quick example code here: https://onlinegdb.com/inTfVSt0a

Now back to the original question, this PR changes 170 files. Would it be possible to split the changes in different PRs so that tracking back changes in the project are easier? Are those 170 files actually changed or is it just a side effect of rebasing?

joluxer commented:
Actually I changed only some files, mainly the Baremetal.ino, the other are a remainder of rebasing and perhaps line endings. I've added the gitattributes, to get rid of the recurring questions from git and editors, which style of line endings should be used. As I'm editing on all three platforms, when doing lots of testing, the gitattributes removed those questions, but it might be, that this touched some more files than expected in the first turn.

joluxer commented:
Ok, I checked my git-log. Those 170 files must come from rebasing, as the only commit, that changes anything in my git-log for this PR is [6d3285c9252304ec6c80bee92fb8ae9107cd6236](https://github.com/thiagoralves/editor/pull/7/commits/6d3285c9252304ec6c80bee92fb8ae9107cd6236) and this touches Barmetal.ino only.

But before things gets worse, I prefer to ask: should I rebase this PR again? Alternatives are to send you the diff/patch, applying the patch here against the master, but then my next PRs might have the same problem. Would you suggest something else? 

Here is a small insight into my git-log:
![Bildschirmfoto_2025-02-09_17-57-55](https://github.com/user-attachments/assets/01a2e4d7-9d0b-4255-99ad-da0f830ababc)

This seems to be such a point, where git makes a lot of work without any advantage of using it. What is your advice, how to proceed, @thiagoralves ?

Regards, JoLuxer

joluxer commented:
I did a rebase on the current editor development and this works out to change way less files. So I'll open this again as soon as the recent editor changes are merged. Then only the single file change in Baremetal.ino should show up.

thiagoralves commented:
Excellent! Thanks @joluxer . Are you going to add any more changes to the other branch? Can I test it all and merge?

joluxer commented:
The other two pull request are ready for test and merge from my perspective.
